packages/sources/canton-functions/src/shared/canton-client.ts

```diff
@@ -0,0 +1,113 @@
+import { Requester } from '@chainlink/external-adapter-framework/util/requester'
+
+export interface CantonClientConfig {
+  AUTH_TOKEN: string
+}
+
+export interface QueryContractByTemplateRequest {
+  templateIds: string[]
+  filter?: string | Record<string, any>
+}
+
+export interface ExerciseChoiceRequest {
+  contractId: string
+  templateId: string
+  choice: string
+  argument: Record<string, any>
+}
+
+export interface Contract {
+  contractId: string
+  templateId: string
+  payload: Record<string, any>
+  signatories: string[]
+  observers: string[]
+  agreementText: string
+  createdAt?: string
+}
+
+export interface ExerciseResult {
+  exerciseResult: any
+  events: any[]
+}
+
+export class CantonClient {
+  private requester: Requester
+  private config: CantonClientConfig
+  private static instance: CantonClient
+
+  constructor(requester: Requester, config: CantonClientConfig) {
+    this.requester = requester
+    this.config = config
+  }
+
+  static getInstance(requester: Requester, config: CantonClientConfig): CantonClient {
+    if (!this.instance) {
+      this.instance = new CantonClient(requester, config)
+    }
+
+    return this.instance
+  }
+
+  /**
+   * Query contracts by template ID with an optional filter
+   */
+  async queryContractsByTemplate(
+    url: string,
+    request: QueryContractByTemplateRequest,
+  ): Promise<Contract[]> {
+    const baseURL = `${url}/v1/query`
+
+    const requestData: any = {
+      templateIds: request.templateIds,
+    }
+
+    if (request.filter) {
```

Comment:
We can either

- Return an error if duplicates are found
- We could also return the full response (all the contracts) and pass this down to the implementing project to handle the return values. This might be preferable